aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1281,6 +1308,118 @@ std::tuple<Tensor, Tensor> _symeig_helper_cuda(const Tensor& self, bool eigenvec
+    scalar_t *work_data = nullptr;
+    ALLOCATE_ARRAY(work_data, scalar_t, lwork);
+    magmaEig<scalar_t>(MagmaNoVec, jobvr, n, self_data, n, wr, wi, NULL, 1, vr_data, ldvr, work_data, lwork, &info);
+    *info_ptr = info;
+  }
+#endif
+}
+
+/*
+ * Internal helper; like eig_cuda but:
+ *   1. assume that self is a square matrix of side "n"
+ *   2. return CPU tensors, which will be copied to GPU memory by the caller
+ */
+static std::tuple<Tensor,Tensor> eig_cuda_helper(const Tensor & self, int64_t n, bool eigenvectors) {
+  // copy self to pinned CPU memory
+  auto self_working_copy = at::empty_strided(
+      {n, n}, // square matrix
+      {1, n}, // column-ordered, as magmaEig expects
+      at::TensorOptions(at::kCPU).dtype(self.dtype()).pinned_memory(true));
+  self_working_copy.copy_(self);
+
+  // tensors holding the results. We use empty_strided to make them column-ordered
+  auto options = self.options().device(at::kCPU).memory_format(LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  auto out_eigvals = at::empty_strided({n, 2}, {1, n}, options);
+  auto out_eigvecs = eigenvectors
+                     ? at::empty_strided({n, n}, {1, n}, options)
+                     : Tensor();
+
+  int64_t info;
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "eig_cuda", [&]{
+    apply_eig<scalar_t>(self_working_copy, eigenvectors, out_eigvals, out_eigvecs, &info);
+  });
+  singleCheckErrors(info, "eig_cuda");
+
+  return std::tuple<Tensor, Tensor>(out_eigvals, out_eigvecs);
+}
+
+std::tuple<Tensor &,Tensor &> eig_cuda_out(Tensor & e, Tensor & v, const Tensor & self, bool eigenvectors) {
+  TORCH_CHECK(self.dim() == 2, "A should be 2 dimensional");
+  squareCheckInputs(self);
+  int64_t n = self.size(-1);
+
+  at::native::resize_output(e, {n, 2});
+  if (eigenvectors)
+      at::native::resize_output(v, self.sizes());
+
+  Tensor cpu_vals, cpu_vecs;
+  std::tie(cpu_vals, cpu_vecs) = eig_cuda_helper(self, n, eigenvectors);
+  e.copy_(cpu_vals);
+  if (eigenvectors)
+      v.copy_(cpu_vecs);
+  return std::tuple<Tensor&, Tensor&>(e, v);
+}
+
+std::tuple<Tensor,Tensor> eig_cuda(const Tensor & self, bool eigenvectors) {
+  TORCH_CHECK(self.dim() == 2, "A should be 2 dimensional");
+  squareCheckInputs(self);
+  int64_t n = self.size(-1);
+
+  // optimization: if self is empty, we can immediately allocate&return empty
+  // GPU tensors, instead of getting empty CPU tensors from eig_cuda_helper
+  // and copying them to GPU
+  if (self.numel() == 0) {
+    auto eigvals = at::empty({n, 2}, self.options());
+    return std::tuple<Tensor, Tensor>(eigvals, at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT));
+  }
+
+  Tensor cpu_vals, cpu_vecs;
+  std::tie(cpu_vals, cpu_vecs) = eig_cuda_helper(self, n, eigenvectors);
+
+  if (eigenvectors) {
+      return std::tuple<Tensor, Tensor>(cpu_vals.to(self.device()), cpu_vecs.to(self.device()));
+  } else {
+      return std::tuple<Tensor, Tensor>(cpu_vals.to(self.device()), at::empty({0}, self.options()));
```

Comment:
Technically I think the empty tensor should use cpu_vecs options but with the device of self, right?

Reply:
true! Although this code has been deleted now :)